src/Lykke.AzureStorage.Test/TableStorageEntity/Metamodel/Providers/AnnotationsBasedMetamodelProviderTests.cs

```diff
@@ -0,0 +1,330 @@
+using Lykke.AzureStorage.Tables.Entity.Serializers;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using Moq;
+
+namespace Lykke.AzureStorage.Test.TableStorageEntity.Metamodel.Providers
+{
+    [TestClass]
+    public class AnnotationsBasedMetamodelProviderTests
+    {
+        #region Mocks
+
+        private class ValueSerializerMock : IStorageValueSerializer
+        {
+            public string Serialize(object value)
+            {
+                throw new NotImplementedException();
+            }
+
+            public object Deserialize(string serialized)
+            {
+                throw new NotImplementedException();
+            }
+        }
+
+        private class OneMoreValueSerializerMock : IStorageValueSerializer
+        {
+            public string Serialize(object value)
+            {
+                throw new NotImplementedException();
+            }
+
+            public object Deserialize(string serialized)
+            {
+                throw new NotImplementedException();
+            }
+        }
+
+        private class AnotherValueSerializerMock : IStorageValueSerializer
+        {
+            public string Serialize(object value)
+            {
+                throw new NotImplementedException();
+            }
+
+            public object Deserialize(string serialized)
+            {
+                throw new NotImplementedException();
+            }
+        }
+
+        [ValueSerializer(typeof(AnnotationsBasedMetamodelProviderTests))]
+        private class SerializableTestTypeWithInvalidSerializerType
+        {
+        }
+
+        [ValueSerializer(typeof(ValueSerializerMock))]
+        private class SerializableTestType
+        {
+        }
+
+        private class DescendantSerializableTestType : SerializableTestType
+        {
+        }
+
+        [ValueSerializer(typeof(OneMoreValueSerializerMock))]
+        private class DescendantSerializableTestTypeWithOverride : SerializableTestType
+        {
+        }
+
+        private class SecondLevelDescendantSerializableTestType : DescendantSerializableTestTypeWithOverride
+        {
+        }
+
+        private class NotSerializableTestType
+        {
+        }
+
+        private class TestTypeWithProperties
+        {
+            [ValueSerializer(typeof(OneMoreValueSerializerMock))]
+            public virtual NotSerializableTestType AnnotatedNotSerializableType { get; set; }
+
+            [ValueSerializer(typeof(OneMoreValueSerializerMock))]
+            public virtual SerializableTestType AnnotatedSerializableType { get; set; }
+
+            public virtual NotSerializableTestType NotAnnotatedNotSerializableType { get; set; }
+
+            public virtual SerializableTestType NotAnnotatedSerializableType { get; set; }
+        }
+
+        private class DescendantTestTypeWithProperties : TestTypeWithProperties
+        {
+        }
+
+        private class DescendatnTestTypeWithOverridenProperties : TestTypeWithProperties
```

Comment:
Typo: Descendatn

Reply:
Yep. Already fixed. Will be updated in the next PR